pkg/tar/stream.go

```diff
@@ -0,0 +1,151 @@
+package tar
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+func StreamFunc(w io.Writer, dir, relativePath string, writeFunc func(f os.FileInfo, shardRelativePath, fullPath string, tw *tar.Writer) error) error {
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+
+	return filepath.Walk(dir, func(path string, f os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Skip adding an entry for the root dir
+		if dir == path && f.IsDir() {
+			return nil
+		}
+
+		// Figure out the the full relative path including any sub-dirs
+		subDir := strings.TrimPrefix(path, dir+string(filepath.Separator))
+		subDir = strings.TrimSuffix(subDir, f.Name())
+
+		return writeFunc(f, filepath.Join(relativePath, subDir), path, tw)
+	})
+}
+
+// Stream writes a tar archive to w, containing the files and sub-directories in the directory dir
+// stored relative to relativePath.
+func Stream(w io.Writer, dir, relativePath string) error {
+	return StreamFunc(w, dir, relativePath, TarFile)
+}
+
+func SinceFilterTarFile(since time.Time) func(f os.FileInfo, shardRelativePath, fullPath string, tw *tar.Writer) error {
+	return func(f os.FileInfo, shardRelativePath, fullPath string, tw *tar.Writer) error {
+		if f.ModTime().After(since) {
+			return TarFile(f, shardRelativePath, fullPath, tw)
+		}
+		return nil
+	}
+}
+
+func TarFile(f os.FileInfo, shardRelativePath, fullPath string, tw *tar.Writer) error {
+	h, err := tar.FileInfoHeader(f, f.Name())
+	if err != nil {
+		return err
+	}
+	h.Name = filepath.Join(shardRelativePath, f.Name())
```

Comment:
This should probably use `filepath.ToSlash` to keep all paths in tar forward slash-based, and then use `filepath.FromSlash` on extraction. Otherwise backups from one system might not be able to be restored onto a system with a different separator.